python/shotgun_fields/bubble_widget.py

```diff
@@ -0,0 +1,495 @@
+        for i in reversed(range(0, len(text))):
+
+            if text[i] == self._OBJECT_REPLACEMENT_CHAR:
+                continue
+
+            cursor.setPosition(i, QtGui.QTextCursor.MoveAnchor)
+            cursor.movePosition(QtGui.QTextCursor.Right, QtGui.QTextCursor.KeepAnchor)
+            cursor.removeSelectedText()
+
+        cursor.endEditBlock()
+
+    def eventFilter(self, object, event):
+        """
+        Attempts to identify clicks on one of the editor's bubble widget's
+        remove button.
+
+        :param object: The observed object.
+        :type object: :class:`~PySide.QtCore.QObject`
+        :param event: The event to filter.
+        :type object: :class:`~PySide.QtCore.Qt.QEvent`
+        :return: True'' if the event was filtered, ''False'' otherwise.
+        """
+        if not isinstance(event, QtGui.QMouseEvent):
+            # only pass on mouse events
+            return False
+
+        # @todo:
+        # can't seem to figure out how to map the viewport position to the
+        # scroll area position. tried all combinations of mapto/from...
+        # resulting to just adding the scroll values. please fix if you can
+        edit_pos = QtCore.QPoint(
+            event.pos().x() + self.horizontalScrollBar().value(),
+            event.pos().y() + self.verticalScrollBar().value()
+        )
+
+        # for mouse events find the actual widget at the position
+        doc = self.document()
+        cursor_pos = doc.documentLayout().hitTest(edit_pos, QtCore.Qt.ExactHit)
+        format = doc.documentLayout().format(cursor_pos)
+        bubble_id = format.property(_BubbleTextObject.BUBBLE_DATA_PROPERTY)
+        bubble = self.get_bubble(bubble_id)
+
+        if bubble is None:
+            self.viewport().setCursor(QtCore.Qt.IBeamCursor)
+            return False
+
+        self.viewport().setCursor(QtCore.Qt.ArrowCursor)
+
+        if event.type() == QtCore.QEvent.MouseButtonPress:
+            # if we are clicking on the button, do so
+            bubble_pos = bubble.mapFromParent(edit_pos)
+            child_widget = bubble.childAt(bubble_pos)
+            if isinstance(child_widget, QtGui.QPushButton):
+                child_widget.click()
+
+        return False
+
+    def get_bubble(self, bubble_id):
+        """
+        Returns a bubble widget based on the supplied id.
+
+        The ``bubble_id`` should correspond to the unique ID returned by the
+        :meth:`.add_bubble` method.
+
+        :param bubble_id: The id of the bubble to retrieve.
+        :return: A bubble widget or ``None`` if not match is found
+        :rtype: :class:`.BubbleWidget`
+        """
+
+        if not bubble_id in self._formats:
+            return None
+
+        text = self.toPlainText()
+        cursor = self.textCursor()
+
+        # loop over each character until a replacement character with a known
+        # format that matches the supplied id is found.
+        for i in range(0, len(text)):
```

Comment:
Minor: `xrange` is technically more memory efficient since it's a generator and the length of the text here isn't known.